### PR TITLE
Updated import to use OSCAL statements property.

### PR DESCRIPTION
**Pull request by omarabed15**
Supercedes: https://github.com/GovReady/govready-q/pull/1218

Updates the component import to create statements based on the "statements" OSCAL JSON property.
Updates the tests to be more robust and test more edge cases (including invalid components and control extensions).
Moved some messages to logs, to reduce UX messages displayed.

### CODE REVIEW

**Comment by woodt** [comment]
So I checked out this branch to do some testing against my work on putting statements into the OSCAL component download.
I think there is a problem in get_control_statement_part():

```
>>> from controls.utilities import get_control_statement_part
>>> get_control_statement_part("at-4_smt.a")
'a'
# that looks correct, part "a"
>>> get_control_statement_part("at-4_smt")
'at-4'
# I think this should return None or an empty string; "at-4" is not a good part identifier
```

**Comment by omarabed15** [comment]
> So I checked out this branch to do some testing against my work on putting statements into the OSCAL component download.
> I think there is a problem in get_control_statement_part():
> 
> ```
> >>> from controls.utilities import get_control_statement_part
> >>> get_control_statement_part("at-4_smt.a")
> 'a'
> # that looks correct, part "a"
> >>> get_control_statement_part("at-4_smt")
> 'at-4'
> # I think this should return None or an empty string; "at-4" is not a good part identifier
> ```

Good point, this was a rudimentary first pass that I forgot to clean up. I'll do that now.

**Comment by omarabed15** [comment]
> > So I checked out this branch to do some testing against my work on putting statements into the OSCAL component download.
> > I think there is a problem in get_control_statement_part():
> > ```
> > >>> from controls.utilities import get_control_statement_part
> > >>> get_control_statement_part("at-4_smt.a")
> > 'a'
> > # that looks correct, part "a"
> > >>> get_control_statement_part("at-4_smt")
> > 'at-4'
> > # I think this should return None or an empty string; "at-4" is not a good part identifier
> > ```
> 
> Good point, this was a rudimentary first pass that I forgot to clean up. I'll do that now.

This should be fixed now.

**Comment by woodt** [comment]
I think there is still a problem with 800-171 style control identifiers:

```
 >>> get_control_statement_part("ac-1_smt")
''
# good!
>>> get_control_statement_part("at-4_smt.b")
'b'
# also good
>>> get_control_statement_part("3.1.1_smt")
'1'
# there is no part in the 3.1.1 identifier, so I think this should be returning an empty string instead.
```

**Comment by omarabed15** [comment]
> I think there is still a problem with 800-171 style control identifiers:
> 
> ```
>  >>> get_control_statement_part("ac-1_smt")
> ''
> # good!
> >>> get_control_statement_part("at-4_smt.b")
> 'b'
> # also good
> >>> get_control_statement_part("3.1.1_smt")
> '1'
> # there is no part in the 3.1.1 identifier, so I think this should be returning an empty string instead.
> ```

Good point. So, to clarify, is the "part" always a single character at the end of control id statement? And should they be validated against existing control id statements in some registry? Right now, it's flexible in the sense that any "part" can be used as long as it fits the pattern, but should I be validating the entire control statement against some registry?

**Comment by woodt** [comment]
Well, a couple of things :-)

(1) Statement ids are supposed to refer to elements in the catalog.  If you look in the 800-53 rev4 catalog, for example (in `controls/data/catalogs`), you'll see lots of statement ids:

```
grep _smt controls/data/catalogs/NIST_SP-800-53_rev4_catalog.json 
                "id": "ac-1_smt",
                    "id": "ac-1_smt.a",
                        "id": "ac-1_smt.a.1",
                        "id": "ac-1_smt.a.2",
                    "id": "ac-1_smt.b",
                        "id": "ac-1_smt.b.1",
                        "id": "ac-1_smt.b.2",
                "id": "ac-2_smt",
                    "id": "ac-2_smt.a",
                    "id": "ac-2_smt.b",
                    "id": "ac-2_smt.c",
                    "id": "ac-2_smt.d",
                    "id": "ac-2_smt.e",
                    "id": "ac-2_smt.f",
                    "id": "ac-2_smt.g",
                    "id": "ac-2_smt.h",
```
etc.,  etc.

So, in theory, incoming statement ids from OSCAL we are importing should match statement ids in the catalog.

(2) In practice, it doesn't look like 800-171 has "parts".  800-53 has parts that tend to be a single letter, but there are some numbered subparts.  (AC-1 b.1, for example.)

I don't know if we need to capture that level of granularity.

Ideally, we would validate a statement id by asserting that it appears in the catalog.  Do we absolutely have to?  Maybe.  I don't think the GovReady UI does any validation, in the sense that I think I can create an AC-1 part (c) statement, even though there is no AC-1(c) in the catalog.

(You and I talked about this idea -- when the catalog JSON gets loaded, we could build a simple index of all the statement ids so that we wouldn't have to go poking around in the hierarchical structure every time we needed to test a statement id.)

**Comment by omarabed15** [comment]
> Well, a couple of things :-)
> 
> (1) Statement ids are supposed to refer to elements in the catalog. If you look in the 800-53 rev4 catalog, for example (in `controls/data/catalogs`), you'll see lots of statement ids:
> 
> ```
> grep _smt controls/data/catalogs/NIST_SP-800-53_rev4_catalog.json 
>                 "id": "ac-1_smt",
>                     "id": "ac-1_smt.a",
>                         "id": "ac-1_smt.a.1",
>                         "id": "ac-1_smt.a.2",
>                     "id": "ac-1_smt.b",
>                         "id": "ac-1_smt.b.1",
>                         "id": "ac-1_smt.b.2",
>                 "id": "ac-2_smt",
>                     "id": "ac-2_smt.a",
>                     "id": "ac-2_smt.b",
>                     "id": "ac-2_smt.c",
>                     "id": "ac-2_smt.d",
>                     "id": "ac-2_smt.e",
>                     "id": "ac-2_smt.f",
>                     "id": "ac-2_smt.g",
>                     "id": "ac-2_smt.h",
> ```
> 
> etc., etc.
> 
> So, in theory, incoming statement ids from OSCAL we are importing should match statement ids in the catalog.
> 
> (2) In practice, it doesn't look like 800-171 has "parts". 800-53 has parts that tend to be a single letter, but there are some numbered subparts. (AC-1 b.1, for example.)
> 
> I don't know if we need to capture that level of granularity.
> 
> Ideally, we would validate a statement id by asserting that it appears in the catalog. Do we absolutely have to? Maybe. I don't think the GovReady UI does any validation, in the sense that I think I can create an AC-1 part (c) statement, even though there is no AC-1(c) in the catalog.
> 
> (You and I talked about this idea -- when the catalog JSON gets loaded, we could build a simple index of all the statement ids so that we wouldn't have to go poking around in the hierarchical structure every time we needed to test a statement id.)

Oh that's right, okay thank you, I need to take another look at the catalog. I'll take a quick look at validating against the catalog. If that gives me much trouble, I'll stay consistent with GovReady's functionality (for now) and not yet validate against the catalog, staying a little flexible.
Based on your (very helpful!) examples, I'm now realizing the part comes after the "_smt.", so I'll refactor to split the control statement id on that string, rather than what I was doing previously.

**Comment by woodt** [comment]
This looks good!

**Review by woodt — APPROVED** [review]

File: controls/utilities.py [hunk 1]
```diff
@@ -78,3 +78,20 @@ def oscalize_catalog_key(catalogkey):
         catalogkey = split_key_list[0] + "-800-" + split_key_list[1]
 
     return catalogkey
+
+def get_control_statement_part(control_stmnt_id):
```
[the remark this answers is not in the document]
Comment (omarabed15) [reply controls/utilities.py:83]: I assume we will call it on that style control identifier. I would think this parser would still work though, because assuming the statement with a part would look like "3.2.1_smt.b", the parser in either case is just looking for the part after "_smt.", which is the same in both situations. Right?
